src/main/java/org/tarantool/jdbc/SQLPreparedStatement.java

```diff
@@ -407,4 +414,73 @@ public void clearBatch() throws SQLException {
         return objects;
     }
 
+    private void ensureLengthLowerBound(long length) throws SQLException {
+        if (length < 0) {
+            throw new SQLException("Stream size cannot be negative", SQLStates.INVALID_PARAMETER_VALUE.getSqlState());
+        }
+    }
+
+    private void ensureLengthUpperBound(long length) throws SQLException {
+        if (length > Integer.MAX_VALUE) {
+            throw new SQLException("Stream size is too large", SQLStates.INVALID_PARAMETER_VALUE.getSqlState());
+        }
+    }
+
+    private void setCharStream(int parameterIndex,
+                               InputStream parameterValue,
+                               long length,
+                               String encoding) throws SQLException {
+        if (length > Integer.MAX_VALUE) {
+            throw new SQLException("Too large stream", SQLStates.INVALID_PARAMETER_VALUE.getSqlState());
+        }
+        try {
+            byte[] bytes = convertToBytes(parameterValue, length);
+            setParameter(parameterIndex, new String(bytes, 0, bytes.length, encoding));
+        } catch (UnsupportedEncodingException e) {
+            throw new SQLException("Unsupported encoding", SQLStates.INVALID_PARAMETER_VALUE.getSqlState(), e);
+        }
+    }
+
+    private void setCharStream(int parameterIndex, Reader reader, long length) throws SQLException {
+        ensureLengthUpperBound(length);
+        try {
+            StringBuilder value = new StringBuilder(4096);
+            char[] buffer = new char[4096];
+            int totalRead = 0;
+            int charsRead;
+            while (totalRead < length &&
+                (charsRead = reader.read(buffer, 0, (int) Math.min(length - totalRead, 4096))) != -1) {
+                value.append(buffer, 0, charsRead);
+                totalRead += charsRead;
+            }
+            setParameter(parameterIndex, value.toString());
+        } catch (IOException e) {
+            throw new SQLException("Cannot read from the reader", SQLStates.INVALID_PARAMETER_VALUE.getSqlState(), e);
+        }
+    }
+
+    private void setBinStream(int parameterIndex,
+                              InputStream parameterValue,
+                              long length) throws SQLException {
+        ensureLengthUpperBound(length);
+        setBytes(parameterIndex, convertToBytes(parameterValue, length));
+    }
+
+    private byte[] convertToBytes(InputStream parameterValue, long length) throws SQLException {
+        try {
+            int bytesRead;
+            int totalRead = 0;
+            byte[] buffer = new byte[4096];
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream(4096);
+            while (totalRead < length &&
+                (bytesRead = parameterValue.read(buffer, 0, (int) Math.min(length - totalRead, 4096))) != -1) {
+                outputStream.write(buffer, 0, bytesRead);
```

Comment:
I think we can write a potentially large request chunk-by-chunk to a socket. To achieve that we need to learn our msgpack encoder to work in a lazy manner: receive a stream as an input parameter and return a stream as a result. Is this a significant amount of work?

Reply:
filed as a separate issue (#219)

Reply:
So let's discuss it in the issue. Okay for this implemetation so.